src/content/post/class3.md

```diff
@@ -0,0 +1,171 @@
+
+Optimization-based attacks generate adversarial examples using gradients obtained through backpropagation. The authors identify obfuscated gradients as an element of many defenses that provides a facade of security by causing the attacker’s gradient descent to fail. Obfuscated gradients come in at least three forms: shattered gradients, stochastic gradients, and vanishing/exploding gradients.
+
+A defense uses **shattered gradients** when it introduces non-differentiable operations, numeric instability, or otherwise causes the attacker’s gradient signal to be incorrect. A defense uses **stochastic gradients** when the inputs to the classifier are randomized or a stochastic classifier is used, resulting in a different gradient each time it is evaluated. **Vanishing/exploding gradients** are issues encountered in training some networks in which the gradient grows or shrinks exponentially during backpropagation. Some defenses involve multiple iterations of neural network evaluation. This can be viewed as evaluating one very deep neural network, which obfuscates the gradient signal by forcing the vanishing/exploding gradient problem.
+
+
+### Detecting Obfuscated Gradients
+The authors propose a number of tests that might help detect when a defense relies on obfuscated gradients.
+
+- Iterative attacks should work better than single-step attacks, since iterative attacks are strictly stronger than single-step attacks.
+- White-box attacks should perform better than black-box attacks, since the black-box threat model is a strict subset of the white-box threat model.
+- Attacks with an unbounded distortion metric (e.g. \\(L_2\\) norm) should find adversarial examples with 100% success.
+- Optimization-based attacks should perform better than brute-force sampling of nearby inputs (sampling within an \\(\epsilon\\)-ball).
+
+These tests may not cover all cases of obfuscated gradients, but they indicate when intuitive properties start to break down. All defenses with obfuscated gradients discussed by the authors fail at least one test.
+
+### Attack Techniques
+The authors introduce attack techniques to address each of the three types of obfuscated gradients.
+
+Shattered gradients can be attacked using a technique called Backward Pass Differentiable Approximation. For a classifier \\(f(\cdot)\\), some defenses make gradients unavailable to attackers by incorporating a non-differentiable preprocessing step \\(g(\cdot)\\), making the secured classifier \\(\hat{f}(\cdot) = f(g(\cdot))\\). Backward Pass Differentiable Approximation estimates gradients by substituting a smooth, differentiable approximation of \\(g\\).
```

Comment:
There's a bug that causes MathJax to render `$\hat{f}(\cdot)$` wrong in some versions of Safari. Is this an issue we need to address? https://github.com/mathjax/MathJax/issues/1737

Reply:
There are some annoying problems with trying to get the markdown mathjax to work correctly...for now, I've gone ahead and posted things, and its probably not worth spending too much effort trying to fix this, so long as people can understand the intent of the equation when it doesn't render correctly.